Purge old DELETEs during live compaction

weggert opened the pull request:
This change is to address Issue 112, 'Old deletes need to be purged
during Compaction'.

It adds the ability to specify a
sirius.log.compaction-max-delete-age-hours configuration property to
cause old Delete events to be purged during live compaction.

The configuration property defaults to a very large value, so by default
Delete events will not be purged.

It also adds an option to WalTool to purge old Deletes, along with
other options for old PUTs.

weggert commented:
Pull request https://github.com/Comcast/sirius/pull/113 addresses this issue.

pcleary00 commented on src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala:
```diff
@@ -317,6 +355,42 @@ object WalTool {
+  }
+
+  private def filterDeletesOlderThan(evt: OrderedEvent, earliestTimestamp: Long): Option[OrderedEvent] = {
+    val request = evt.request
```
Don't really need the local variable here, can just do `evt.request match`

clinedome-work commented on src/main/scala/com/comcast/xfinity/sirius/api/SiriusConfiguration.scala:
```diff
@@ -195,6 +195,12 @@ object SiriusConfiguration {
 
   /**
+   * The maximum age for a Delete in hours that remains after compaction.  Deletes
+   * that are older than this age will be purged during compaction.
```
Can you add text about the default value here?

weggert commented:
I added documentation to the purge deletes config item, and removed the unnecessary variable from WalTool.

clinedome-work commented on src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala:
```diff
@@ -15,15 +15,28 @@
 
+  def apply(siriusConfig: SiriusConfiguration) : SegmentedCompactor = {
+
+    val maxDeleteAgeHours = siriusConfig.getProp(SiriusConfiguration.COMPACTION_MAX_DELETE_AGE_HOURS, Integer.MAX_VALUE) // a really long time by default
```
Please remove the trailing comment, I think it goes over the 120 char max width. It can either go on the line above or it can just be removed, I think it's fine either way.

clinedome-work commented on src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala:
```diff
@@ -15,15 +15,28 @@
+  }
+}
+
+class SegmentedCompactor(maxDeleteAgeMillis: Long) {
```
This can probably made at least package private. Outside of tests, the apply() method should be used to instantiate this class.

clinedome-work commented on src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala:
```diff
@@ -92,17 +105,19 @@ object SegmentedCompactor {
-    // XXX if we want to remove old deletes from the log...
-    // val currentTime = System.currentTimeMillis
+    val currentTime = System.currentTimeMillis
+    val earliestDelete = if (maxDeleteAgeMillis > currentTime) 0L else currentTime - maxDeleteAgeMillis
```
This could use a comment. I guess none of the updates are likely to come from the 60's, so we're probably not missing any deletes with this, but it's a little confusing, and I'm not sure it makes this any safer/simpler/easier. We could probably remove the `if` statement and just keep it simply `currentTime - maxDeleteAgeMillis`

clinedome-work commented on src/main/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactor.scala:
```diff
@@ -92,17 +105,19 @@ object SegmentedCompactor {
+
     source.foreach {
-      // case OrderedEvent(_, timestamp, Delete(key)) if (currentTime - timestamp) > MAX_DELETE_AGE => // nothing
+      case OrderedEvent(_, timestamp, Delete(key)) if timestamp < earliestDelete => // nothing
```
I do like using `earliestDelete` so we don't have to see that comparison happen down here.

clinedome-work commented on src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala:
```diff
@@ -232,11 +249,34 @@ class SegmentedCompactorTest extends NiceTest with BeforeAndAfterAll {
+
+    val compacted = makeSegment(compactionMap(first))
+
+    assert("2" === listEvents(compacted))
```
This looks good.

clinedome-work commented:
All in all, this looks great. A few minor comments, but with those addressed, I'd be +1. Thanks @weggert !

weggert commented:
Per the feedback I pushed up changes to make the SegmentedCompactor private, remove the unnecessary comment, and simplify the Delete purge logic in the SegmentedCompactor.

clinedome-work commented:
Awesome. Looks great. :+1:

tbarker9comcast commented on src/main/scala/com/comcast/xfinity/sirius/tool/WalTool.scala:
```diff
@@ -81,12 +81,24 @@ object WalTool {
     Console.err.println("       Same as key-filter, except the resulting UberStore contains all")
     Console.err.println("       OrderedEvents not matching regexp")
     Console.err.println()
+    Console.err.println("   key-filter-deletes-not-older-then <earliestMillis> <inWalDir> <outWalDir>")
```
`s/then/than`

tbarker9comcast commented on src/test/scala/com/comcast/xfinity/sirius/uberstore/segmented/SegmentedCompactorTest.scala:
```diff
@@ -232,11 +249,34 @@ class SegmentedCompactorTest extends NiceTest with BeforeAndAfterAll {
     }
   }
 
+  it("should remove all delete events in appropriate segments that are older than COMPACTION_MAX_DELETE_AGE_HOURS") {
```
Since we are considering the corner case in the code, can you create a test when the age is greater than the current time (ie nothing is deleted)?

tbarker9comcast commented:
Had some minor comments, but once they are addressed this has my :+1:. This looks great to me.

weggert commented:
I fixed the typo in the WalTool documentation, and I added a new test for the SegmentedCompactor Delete purge logic to make sure it works as expected when the max-time is bigger than the current time.

There is also a minor change for when getting the property to change siriusConfig.getProp(...) to siriusConfig.getInt(...) to avoid ClassCastExceptions if the property value is set with a Long.

weggert commented:
Thanks for the reviews.  :-)  I squashed the updates into the original commit

clinedome-work commented:
:+1:
